app/src/main/java/de/westnordost/streetcomplete/map/LocationAwareMapFragment.kt

```diff
@@ -106,83 +131,145 @@ open class LocationAwareMapFragment : MapFragment() {
     }
 
     private fun onLocationChanged(location: Location) {
-        this.displayedLocation = location
+        displayedLocation = location
         locationMapComponent?.location = location
+        addTrackLocation(location)
         compass.setLocation(location)
         centerCurrentPositionIfFollowing()
-        listener?.onLocationDidChange()
+        listener?.onDisplayedLocationDidChange()
     }
 
-    /* --------------------------------- Rotation tracking -------------------------------------- */
-
-    private fun onCompassRotationChanged(rot: Float, tilt: Float) {
-        compassRotation = rot * 180 / PI
-        locationMapComponent?.rotation = compassRotation
+    private fun addTrackLocation(location: Location) {
+        // ignore if too imprecise
+        if (location.accuracy > MIN_TRACK_ACCURACY) return
+        val lastLocation = tracks.last().lastOrNull()
 
-        if (isCompassMode) {
-            viewDirection =
-                if (viewDirection == null) -rot
-                else smoothenAngle(-rot, viewDirection ?: 0f, 0.05f)
+        // create new track if last position too old
+        if (lastLocation != null) {
+            if ((displayedLocation?.time ?: 0) - lastLocation.time > MAX_TIME_BETWEEN_LOCATIONS) {
+                tracks.add(ArrayList())
+                tracksMapComponent?.startNewTrack()
+            }
+        }
 
-            controller?.updateCameraPosition { rotation = viewDirection }
+        tracks.last().add(location)
+        // delay update by 600 ms because the animation to the new location takes that long
+        lifecycleScope.launch {
+            delay(600)
+            tracksMapComponent?.addToCurrentTrack(location)
         }
     }
 
-    private fun smoothenAngle( newValue: Float, oldValue: Float, factor: Float): Float {
-        var delta = newValue - oldValue
-        while (delta > +PI) delta -= 2 * PI.toFloat()
-        while (delta < -PI) delta += 2 * PI.toFloat()
-        return oldValue + factor * delta
+    /* --------------------------------- Rotation tracking -------------------------------------- */
+
+    private fun onCompassRotationChanged(rot: Float, tilt: Float) {
+        locationMapComponent?.rotation = rot * 180 / PI
     }
 
     /* -------------------------------- Save and Restore State ---------------------------------- */
 
     private fun restoreMapState() {
         val prefs = activity?.getPreferences(Activity.MODE_PRIVATE) ?: return
         isFollowingPosition = prefs.getBoolean(PREF_FOLLOWING, true)
-        isCompassMode = prefs.getBoolean(PREF_COMPASS_MODE, false)
+        isNavigationMode = prefs.getBoolean(PREF_NAVIGATION_MODE, false)
     }
 
     private fun saveMapState() {
         activity?.getPreferences(Activity.MODE_PRIVATE)?.edit {
             putBoolean(PREF_FOLLOWING, isFollowingPosition)
-            putBoolean(PREF_COMPASS_MODE, isCompassMode)
+            putBoolean(PREF_NAVIGATION_MODE, isNavigationMode)
```

Comment:
Maybe an int (or string) here, so if there is ever a 3rd mode, you don't need to burn the pref key? Using an enum, it should be ~1 additional line of code.

Reply:
(e.g. perhaps one of the ideas from https://github.com/streetcomplete/StreetComplete/issues/2191)

Reply:
Along these lines
```diff
+ enum class NAVIGATION_MODE(val isEnabled: Boolean) { TOP_DOWN(false), ANGLED(true) }
- private var _isNavigationMode: Boolean = false
+ private var navigationMode: NAVIGATION_MODE = TOP_DOWN
  var isNavigationMode: Boolean
      set(value) {
-         if (_isNavigationMode != value && !value) {
+         if (navigationMode.isEnabled != value && !value) {
              updateCameraPosition(300) { tilt = 0f }
          }
-         _isNavigationMode = value
+         navigationMode = if (value) NAVIGATION_MODE.TOP_DOWN else NAVIGATION_MODE.ANGLED
      }
-     get() = _isNavigationMode
+     get() = navigationMode == NAVIGATION_MODE.ANGLED
```
```suggestion
            putInt(PREF_NAVIGATION_MODE, navigationMode.ordinal)
```

Reply:
Yes, I know what you meant. I don't see the sense in implementing something that isn't used though.

What I changed right now is the key name of the pref to the old key, because it is conceptionally the same as the "compass mode".

You seem to hint at that you'd like to a top-down navigation mode. I actually decreased the tilt angle for the "new" navigation mode now a bit. The gps location button is now a two-state button (follow me <-> navigation mode). It would be too confusing to make it a three-state button.